Don't drop UIDs in GlobalIndexUidAggregator

brianloss opened the pull request:
* Ensure that if there is a positive (including 0) count for an incoming
  protocol buffer that we handle UIDs in the REMOVEDUID list as well as
  those in the UID list. This scenario is possible if a partial major
  compaction combines files adding UID(s) to the UID list and adding
  other/different UID(s) to the REMOVEDUID list (where the adds of the
  UIDs being removed are in a different file that isn't involved in the
  major compaction).

* Switch logger to use slf4j and placeholders in the log messages.

* Add test cases for new functionality, and apply IDE clean up.

Fixes #1173

jschmidt10 commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -150,18 +144,21 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                // a partial major compaction and combined a protocol buffer having a positive
+                // count of 1 and a negative count of 1 (leaving one UID in the UID list and one
+                // UID in the REMOVEDUID list.
+                if (delta >= 0) {
```
Nice catch. I'm wondering if this could even happen with numbers less than 0 - like if one RFile contains an insert of uid1 and another RFile has a removeUidList with uid2 and uid3 in it.

brianloss replied:
@jschmidt10 Good point. Maybe the same action needs to be taken regardless of the count in the incoming protocol buffer -- add the UIDs in the add list, and remove the ones in the remove list. I also have some stuff in here that changes behavior w.r.t timestamps (I'm assuming timestamp matters but I know by default datawave truncates timestamp to the day boundary so I'll need to guard that behavior change with an option). I'll take a look at all that at some point and push some updates.

jschmidt10 replied:
Sounds good. Thanks!

javease commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -138,7 +132,7 @@ public Value aggregate() {
     @Override
     public Value reduce(Key key, Iterator<Value> iter) {
         if (log.isTraceEnabled())
```
similar to elsewhere, could remove the if wrapping the log statement

javease commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -99,8 +95,7 @@ public Value aggregate() {
-                if (log.isDebugEnabled())
-                    log.debug("Adding released UIDS");
+                log.debug("Adding released UIDS");
                 uids.addAll(releasedUids);
```
I realize this isn't part of your change but couldn't this result in uids exceeding the max threshold?

javease commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -150,18 +144,21 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                // UID in the REMOVEDUID list.
+                if (delta >= 0) {
                     
                     for (String uid : v.getQUARANTINEUIDList()) {
```
I'm not fulling understanding the construction of this protobuf. For the case where delta is non-negative, we appear to be treating the QUARANTINEUIDList as "released UIDs" which we'll add back into the list of uids in the aggregator call. For negative delta, this is treated as a quarantine list not to be included in the uids. Is this a correct understanding?

brianloss replied:
Yeah, I'm not really sure about the quarantined/released lists. Maybe @phrocker can comment? Think he knows...

brianloss replied:
It does look like maybe the intent was for the QUARANTINEDUID list to be dual-purpose... You quarantine a UID by doing an insert with a count of -1 and the UID in the QUARANTINEDUID list. You release it by doing an insert with a count of 1 and the UID in the QUARANTINEDUID list. However, with partial major compactions, that approach won't work (or I'm just wrong about how the feature works) since the -1 count PB with the uid in the QUARANTINEDUID list could get combined with a PB having, say, a count of 2 and two other uids in the UID list. The result would be a PB with a count of 1, 2 uids in the UID list, and 1 in the QUARANTINEDUID list. The next time that PB is read in (for any of a scan, partial major compaction, or full major compaction) the delta will be positive and the quarantined UID will be released even though that probably wasn't the intent.

phrocker replied:
hmm, my recollection is that all keys would be propagated until a full major C, so that even if the partial major occurred all keys would exist until that FullMajorC was launched periodically.

brianloss replied:
In the untouched version of this, the uids in the QUARANTINEDUID list are released when the delta is positive (see [here](https://github.com/NationalSecurityAgency/datawave/blob/master/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java#L166)) and uids in the QUARANTINEDUID list are carried forward as quarantined when the delta is negative (see [here](https://github.com/NationalSecurityAgency/datawave/blob/master/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java#L204)).

This seems like a problem for partial major compactions. In case 1, say two PBs are combined. One has a count of -1 and uid1 in the QUARANTINEDUID list. The other PB has a count of 2 and uid2 and uid3 in the UID list. The combined PB will have a count of 1, uid2 and uid3 in the UID list, and uid1 in the QUARANTINEDUID list. Then, next time there's a scan or another compaction, the incoming delta for this new PB will be positive and the quarantined uid1 will be released. For case 2, say a previously quarantined uid1 is released so there's a PB with a count of 1 and uid1 in the QUARANTINEDUID list. Say that PB gets combined with another during a partial major compaction where the other PB has a count of -2 with uid2 and uid3 in the REMOVEDUID list. The final PB will have a count of -1, uid1 in the UID list, and uid2 and uid3 in the REMOVEDUID list. Now when that PB eventually gets combined with the one that originally quarantined uid1, uid1 will be re-quarantined.

phrocker replied:
Yep. I thought there being checks for testing if  partial, not doing any combining of keys at all, leaving scan and full major c to do the work as a safeguard....then periodic full majors doing the combining and reduction of keys...but again, my knowledge is vestigial.

brianloss replied:
Perhaps the way to handle this is to modify the protobuf to add a RELEASEDUID list as well. When the intent is to release a previously quarantined UID, insert a PB with a count of 1 and the uid in the new RELEASEDUID list. Then the action to take in the code is clear and it will work with partial major compactions. We could add a special case to release a UID if the incoming PB contains only uids in the QUARANTINEDUID list and not any of the other lists. That would catch previously released quarantined uids that were never involved in a partial major compaction. For those that were involved in a partial majc, I don't think one can make any assumptions about uids in the QUARANTINEDUID list. @ivakegg, any thoughts? I have some local changes to attempt this and also fix the case @jschmidt10  identified. I'll go ahead and push those, and we can just drop the new commit if people don't like that approach.

brianloss replied:
@javease, I believe my latest change should address your question and also resolve the discussion in this thread.

brianloss replied:
The QUARANTINEDUID list has been removed, so this discussion is OBE. Resolving...

javease commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -172,48 +169,49 @@ public Value reduce(Key key, Iterator<Value> iter) {
                             
                             // add the UID iff we are under our MAX
                             if (uids.size() < maxUids)
                                 uids.add(uid);
```
could add a break here if condition is false since we won't be adding any entries in the for loop once maxUids is reached.

javease commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -172,48 +169,49 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                } else if (!seenIgnore) {
                     
                     // so that we can perform the decrement
                     for (String uid : v.getREMOVEDUIDList()) {
```
again, not fully understanding our use of this protobuf, but in the above block we don't appear to care about the REMOVEDUIDList when propogate is false but here we're updating uidsToRemove regardless of the propagate setting. Does that make sense?

brianloss replied:
If the propogate [sic] var is false, then removals aren't passed along (basically propogate is false for a full major compaction only). So yeah, we should probably be checking the propogate var here too. The seenIgnore flag gets set when there's a protocol buffer whose count exceeds the max. In that case, there's nothing in either the UID or REMOVEDUID list, so we should probably be checking that in the add case as well. I'm not an expert on this class and was just trying to fix a specific error that I encountered. I'm hesitant to make changes that are too sweeping, but if there are other experts on this class who want to weigh in and comment to ensure I haven't broken anything, that would be good. Errors in this class could have some pretty serious negative consequences.

ivakegg commented:
Would it be possible to redo this against the integration branch?  Sorry for the hassle but accumulo-2.0 is somewhat delayed and I would like to get this one out there....

brianloss commented:
> Would it be possible to redo this against the integration branch? Sorry for the hassle but accumulo-2.0 is somewhat delayed and I would like to get this one out there....

Sure, I'll take a look a rebasing it this week. I was in the process of reworking and need to write more tests, but then got busy on other tasks. Hopefully I can spend some time writing more tests this week.

brianloss commented:
> Would it be possible to redo this against the integration branch? Sorry for the hassle but accumulo-2.0 is somewhat delayed and I would like to get this one out there....

@ivakegg, I rebased my branch off integration, force pushed, and updated the PR target to integration. I also added another commit with new changes based on prior discussions in the PR. I think this needs more tests to confirm behavior, but wanted to get eyes on the code changes since it's a complicated class, and and error could have a big impact on a production system.

phrocker commented on warehouse/ingest-core/src/main/java/datawave/iterators/PropogatingIterator.java:
```diff
@@ -303,7 +305,10 @@ public boolean validateOptions(Map<String,String> options) {
+        // Don't propagate for either scan or full major compaction. In either case, the aggregated result has combined
+        // all existing values for a key so we don't need to propagate temporary state that is only used to combine
+        // partial results with new info.
+        shouldPropogate = !(env.getIteratorScope() == IteratorScope.majc && env.isFullMajorCompaction()) && !(env.getIteratorScope() == IteratorScope.scan);
```
I think this change makes it much more explicit which I think is important here.

jschmidt10 commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -240,32 +288,47 @@ public void reset() {
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        super.init(source, options, env);
+        if (options.containsKey(TIMESTAMPS_IGNORED)) {
```
I hit an issue with trying to add options to GlobalIndexUidAggregator - since we're configuring it as an aggregator option on a TotalAggregatingIterator - the `GlobalIndexUidAggregator.init()` method is never actually invoked. Also, the TotalAggregatingIterator treats every option as a column so it'll treat `timestampsIgnored` as the column and try to instantiate a class of `true` for instance.

I also tried using GlobalIndexUidAggregator as a top-level iterator but there is an issue with that as the `propogateKey()` and `reset()` methods are invoked by TotalAggregatingIterator (via PropogatingIterator) - if you use GlobalIndexUidAggregator directly on the table, only `reduce()` gets invoked by Accumulo. 

I had considered maybe adding a hard-coded option to use as a flag to delegate options down into the aggregator like maybe `table.iterator.majc.UidAgg.opt.*.aggropt.timestampsIgnored`. I ended up just using a subclass of GlobalIndexUidAggregator instead though.

brianloss replied:
@jschmidt10, PR #1091 would allow the options to be passed to aggregators via TotalAggregatingIterator. To use GlobalIndexUidAggregator as a Combiner, other work would probably be required.

jschmidt10 replied:
Great! Thanks.

brianloss replied:
That PR also allows the max UID count to be configured. I can roll just the changes that allow the options to be passed along to this PR if that will make it easier to test anything. (Or, I can make that a separate PR.)

jschmidt10 replied:
I think we're set. With ignoreTimestamp being the default, we don't need the option right away.

matthpeterson commented on warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java:
```diff
@@ -224,8 +225,353 @@ public void testCount() throws Exception {
+        assertEquals(1, resultList.getUIDCount());
+        
+    }
+    
```
Thanks for tackling this worthwhile effort to correct the inaccuracies with the aggregator!

The test below adds the same UID twice, highlighting the possible inconsistencies between getCount and the size of getUIDList.

Note that the test fails both before and after this PR's changes.  It's just highlighting another inaccuracy with the aggregator.

```
    @Test
    public void testAddUIDTwice() throws Exception {
        GlobalIndexUidAggregator localAgg = new GlobalIndexUidAggregator();
        IteratorSetting is = new IteratorSetting(19, "test", GlobalIndexUidAggregator.class);
        GlobalIndexUidAggregator.setCombineAllColumns(is, true);
        localAgg.validateOptions(is.getOptions());

        UUID uuid1 = UUID.randomUUID();
        ArrayList<Value> values = Lists.newArrayList();
        Builder b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        Value result = localAgg.reduce(new Key("key"), values.iterator());
        Uid.List resultList = Uid.List.parseFrom(result.get());        
        assertEquals(1, resultList.getUIDCount());
        assertTrue(resultList.getUIDList().contains(uuid1.toString()));
        assertEquals(1, resultList.getUIDList().size());
        assertEquals(1, resultList.getCOUNT()); // Fails here getCount()==2
    }
```

There are use cases in which the same UID will be added multiple times so it's worth considering the behavior.  

The test above highlights how the aggregator will have an inconsistency between its count and list size.  The test below shows how the aggregator will assess that the max was reached without first checking if the UID is unique or already in the UIDList.  

It seems that the same UID added three times (while IGNORE == false) should result in a count of one and a UID list of size one.

```
    @Test
    public void testAddUIDThrice() throws Exception {
// lowered max to show problem more easily
        GlobalIndexUidAggregator localAgg = new GlobalIndexUidAggregator(2);
        IteratorSetting is = new IteratorSetting(19, "test", GlobalIndexUidAggregator.class);
        GlobalIndexUidAggregator.setCombineAllColumns(is, true);
        localAgg.validateOptions(is.getOptions());

        UUID uuid1 = UUID.randomUUID();
        ArrayList<Value> values = Lists.newArrayList();
        Builder b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        Value result = localAgg.reduce(new Key("key"), values.iterator());
        Uid.List resultList = Uid.List.parseFrom(result.get());
        assertEquals(1, resultList.getUIDCount()); // Fails getUIDCount == 0
        assertTrue(resultList.getUIDList().contains(uuid1.toString()));
        assertEquals(1, resultList.getUIDList().size());
        assertEquals(1, resultList.getCOUNT());
    }
```

Perhaps the logic should not depend on count unless seenIgnore is true.  The size of the UID list would be a more accurate count to use, when available.

Note that before the UIDs are put into the protobuf List they are deduplicated in a HashSet within the aggregator.

Maybe this issue is outside the scope of this ticket.

brianloss replied:
Sure, these are good points. I originally wanted only to correct the UIDs getting dropped, and was trying to change as little as possible to make that happen. Then more things cropped up and that lead to adding the released UID list and also re-working the inner logic. It's not unreasonable to try to solve issues with duplicate UIDs too. It gets tricky if you are trying to combine 2 protocol buffers that are near their limit (and have a large max UID count). Do you combine the two sets and have a possibly huge (albeit temporary list)? Do you compute an intersection look at the total count (which may save memory but you also have to do it every time and have the performance hit)?

brianloss replied:
@matthpeterson it turns out there's actually a test that verifies that duplicate UIDs are counted! See [here](https://github.com/NationalSecurityAgency/datawave/blob/integration/warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java#L226).

brianloss replied:
@matthpeterson and @jschmidt10, I have pushed changes to attempt to address duplicate UIDs (and also get rid of the quarantine/release lists). It's still a tricky class to get the behavior right, but hopefully this is an improvement. Have a look (especially what's going on with counts in the code and tests) and let me know what you think.

matthpeterson replied:
Interesting test.  I see now that a lot of the other tests also assert a discrepancy between getCount and getUIDCount().

If a system ever re-ingests the same data and then removes it, the existing logic will leave "dangling" index entries - even for a list that has never hit the max.  I'd rather the logic be changed to be as accurate as possible.  I don't know why we wanted to deliberately have inconsistencies between the count and UIDList sizes except to accommodate the quarantine / release logic.  Now that the logic is being updated to rely on the two separate lists for quarantine and release, I don't think we need to try to conflate the meaning of count.

> It gets tricky if you are trying to combine 2 protocol buffers that are near their limit (and have a large max UID count). Do you combine the two sets and have a possibly huge (albeit temporary list)? Do you compute an intersection look at the total count (which may save memory but you also have to do it every time and have the performance hit)?

The change below seems to correct the discrepancies between getCount() and getUIDCount() and make the two unit tests from my prior comment pass.
```
--- a/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java
+++ b/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java
@@ -87,9 +87,9 @@ public class GlobalIndexUidAggregator extends PropogatingCombiner {

         // as a backup, we remove the intersection of the UID sets

-        builder.setCOUNT(count);
+        builder.setCOUNT(seenIgnore ? count : uids.size());

-        if (seenIgnore || count > maxUids) {
+        if (seenIgnore || uids.size() > maxUids) {
```

This change would require making the getCOUNT and getUIDCount assertions match for the test you mentioned and a few others.
```
@@ -224,13 +224,81 @@ public class GlobalIndexUidAggregatorTest {
         Value result = agg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());

-        assertEquals(5, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
```
and
```
@@ -267,7 +335,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(1, resultList.getCOUNT());
+        assertEquals(2, resultList.getCOUNT());
         assertEquals(2, resultList.getUIDCount());
         assertEquals(2, resultList.getUIDList().size());
         assertEquals(0, resultList.getREMOVEDUIDList().size());
@@ -318,7 +386,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(-2, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
         assertEquals(1, resultList.getUIDCount());
         assertEquals(1, resultList.getUIDList().size());
         assertEquals(3, resultList.getREMOVEDUIDCount());
@@ -476,7 +544,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(0, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
         assertEquals(1, resultList.getUIDCount());
         assertEquals(1, resultList.getUIDList().size());
         assertEquals(1, resultList.getREMOVEDUIDList().size());
@@ -543,7 +611,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(1, resultList.getCOUNT());
+        assertEquals(2, resultList.getCOUNT());
```

brianloss replied:
> The change below seems to correct the discrepancies between getCount() and getUIDCount() and make the two unit tests from my prior comment pass.

I had already made a similar change in my latest commit. I think I must have pushed right as you were typing this up @matthpeterson. Have a look at the latest. It's a bit more of a refactoring since the quarantine list was also removed. It might be easier at this point to look at the raw file vs the diffs.

brianloss replied:
I should add that I looked at enforcing the constraint that `getCOUNT() == (getUIDCount() - getREMOVEDUIDCount())` but found this to not be possible. When we go through the UID list to add them to the in-memory set, we first check if the UID is in the uidsToRemove set and take no action if that's the case. One might suggest removing the uid from uidsToRemove at that point since we've encountered the "add" that goes with the remove. However, it would not be safe to remove the uid from uidsToRemove since we don't yet know if there are any other adds for that uid (from other entries that we haven't seen yet).

matthpeterson commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -150,69 +155,113 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                    // Add any quarantined UIDs to the internal list. If we're paying attention to timestamps, then we
+                    // don't want to quarantine any uid that we've already seen (i.e., with a newer timestamp) that has
+                    // been released.
+                    for (String uid : v.getQUARANTINEUIDList()) {
```
I'm unaware of any uses of the quarantine and release capabilities.  Perhaps it would be best to deprecate and drop support, for the sake of simplifying this complicated and critical aggregator.  Thoughts?

brianloss replied:
It would make the logic simpler, for sure.

brianloss replied:
I will try removing that, and covering the duplicate UID cases.

brianloss commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -150,69 +155,113 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                        uids.remove(uid);
+                }
+                
+                /*
+                 * This is added for backwards compatibility. The removal list was added to ensure that removals are propagated across compactions. In the case
+                 * where compactions did not occur, and the indices are converted into the newer protobuf, we must use the UID list to maintain removals for
+                 * deltas less than 0
+                 */
+                if (delta < 0 && v.getREMOVEDUIDList().isEmpty()) {
+                    for (String uid : v.getUIDList()) {
+                        // add to uidsToRemove, and decrement count if the uid is in UIDS
+                        uidsToRemove.add(uid);
+                        uids.remove(uid);
+                    }
+                }
```
This block was effectively moved from an earlier version. @matthpeterson, are you aware of any cases where this is actually needed? I'm wondering if this can go too?

matthpeterson replied:
I'm thinking that this backwards-compatibility code has been in place for so long now that any system having compacted its indexes since then would no longer need it.

brianloss replied:
Ok, I'll push another commit that removes it.

matthpeterson commented on warehouse/core/src/main/java/datawave/ingest/protobuf/Uid.java:
```diff
@@ -1084,17 +895,15 @@ public Builder addQUARANTINEUIDBytes(com.google.protobuf.ByteString value) {
-                                        "QUARANTINEUID",});
+                                internal_static_datawave_ingest_protobuf_List_descriptor, new java.lang.String[] {"IGNORE", "COUNT", "UID", "REMOVEDUID",});
                 return null;
             }
```
What happens if the new Uid protobuf (with QUARANTINEUID no longer present) is used to deserialized a legacy serialized Uid protobuf (that had the QUARANTINEUID in its schema, but unpopulated in practice)?

brianloss replied:
I thought protocol buffers would just ignore unknown fields, but we should verify.

matthpeterson replied:
@jschmidt10 verified that your understanding is correct.

matthpeterson commented:
https://github.com/brianloss/datawave/pull/6 was created with suggested changes to this PR.  If approved and merged, it will modify this PR's source branch and require another round of approvals here.
